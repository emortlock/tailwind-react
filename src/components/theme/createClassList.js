import merge from 'lodash.merge'
import defaultTheme from './defaultTheme'

const mapObject = (obj, prefix) =>
  Object.keys(obj).map(key => `${prefix}-${obj[key]}`)

const mapArray = (arr, prefix) => arr.map(value => `${prefix}-${value}`)

const reduceMultiplePrefixes = (obj, prefixes) =>
  Object.keys(obj).reduce(
    (arr, key) =>
      typeof obj[key] === 'object'
        ? [...arr, ...reduceMultiplePrefixes(obj[key], prefixes)]
        : [...arr, ...prefixes.map(prefix => `${prefix}-${obj[key]}`)],
    [],
  )

export default (projectTheme = {}) => {
  const theme = merge(defaultTheme, projectTheme)

  return [
    theme.radius,
    ...reduceMultiplePrefixes(theme.spacing, [
      'mb',
      'mr',
      '-ml',
      '-mr',
      'p-',
      'pl-',
      'px-',
      'py-',
    ]),
    ...mapObject(theme.container, 'max-w'),
    ...mapArray(theme.text.size.body, 'text'),
    ...mapArray(theme.text.size.title, 'text'),
    ...mapObject(theme.text.family, 'font'),
    ...reduceMultiplePrefixes(theme.brandColors, [
      'background-color',
      'border-color',
      'text',
    ]),
    ...reduceMultiplePrefixes(theme.textColors, [
      'background-color',
      'border-color',
      'text',
    ]),
  ].filter((className, index, self) => self.indexOf(className) === index)
}
