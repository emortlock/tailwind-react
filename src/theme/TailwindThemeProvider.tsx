import { Theme } from './TailwindTheme'

import React from 'react'
import merge from 'lodash.merge'

import { defaultTheme } from './defaultTheme'
import { TailwindTheme } from './TailwindTheme'

export const TailwindThemeProvider = ({
  theme = {},
  children,
}: {
  theme: Partial<Theme>
  children: React.ReactNode
}) => {
  const mergedTheme: Theme = merge({}, defaultTheme, theme)

  return (
    <TailwindTheme.Provider value={mergedTheme}>
      {children}
    </TailwindTheme.Provider>
  )
}
