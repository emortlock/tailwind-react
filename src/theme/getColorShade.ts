const shades = ['100', '200', '300', '400', '500', '600', '700', '800', '900']

export const getColorShade = (
  color?: string | string[],
  offset: string | number = 1,
): string => {
  if (!color) return ''

  let currentColor: string | string[] = Array.isArray(color) ? color[0] : color

  if (offset === 0) return currentColor

  currentColor =
    currentColor === 'white' ? ['gray', '100'] : currentColor.split('-')
  let shadeOffset = offset

  if (color === 'white') {
    if (shadeOffset < 1) return color
    if (shadeOffset === 1) return 'gray-100'

    currentColor = ['gray', '100']
    if (typeof offset === 'number') shadeOffset = offset - 1
  }

  if (currentColor.length === 1) {
    currentColor.push('400')
  }

  if (typeof shadeOffset === 'string') {
    return `${currentColor[0]}-${shadeOffset}`
  }

  const shadeIndex = Math.min(
    Math.max(shades.indexOf(currentColor[1]) + shadeOffset, 0),
    shades.length - 1,
  )
  const newShade = shades[shadeIndex]

  return `${currentColor[0]}-${newShade}`
}
