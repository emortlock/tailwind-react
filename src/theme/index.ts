export * from './defaultTheme'
export * from './TailwindTheme'
export * from './TailwindThemeProvider'
export * from './withTheme'
export * from './useThemeValue'
export * from './getColorShade'
