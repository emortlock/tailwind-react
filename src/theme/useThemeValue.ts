import { tailwindPropToClassName } from 'tailwind-react-primitives'

export const useThemeValue = (
  prefix: string,
  value: unknown,
  userClassNames = '',
) => {
  if (userClassNames.includes(prefix)) {
    return false
  }

  return tailwindPropToClassName(prefix, value)
}
