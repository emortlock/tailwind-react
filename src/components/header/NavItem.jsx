import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { BaseComponent } from '../tailwind'

const NavItem = ({
  theme,
  is,
  children,
  header: { style },
  active,
  ...rest
}) => {
  const textColor = style.text || theme.textColors.on.primary
  const bgColor = style.bg || theme.brandColors.primary

  return (
    <BaseComponent
      is={is}
      focusable
      text={!active ? style.text : style.bg}
      bg={active ? textColor : undefined}
      bg-hocus={textColor}
      text-hocus={bgColor}
      p={{ x: theme.spacing.md, y: theme.spacing.sm }}
      m={{ t: theme.spacing.sm }}
      m-lg={{ t: 0, r: theme.spacing.sm }}
      rounded={theme.radius}
      noUnderline
      block
      aria-current={active ? 'page' : undefined}
      {...rest}
    >
      {children}
    </BaseComponent>
  )
}

NavItem.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  header: PropTypes.shape({
    style: PropTypes.object.isRequired,
  }),
  active: PropTypes.bool,
}

NavItem.defaultProps = {
  is: 'a',
  children: undefined,
  header: { style: {} },
  active: false,
}

export { NavItem as component }
export default withTheme(NavItem)
