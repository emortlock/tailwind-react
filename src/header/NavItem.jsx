import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { withTheme } from '../theme'
import { BaseComponent } from '../tailwind'

const NavItem = ({
  theme,
  is,
  children,
  className,
  header: { style },
  active,
  ...rest
}) => (
  <BaseComponent
    is={is}
    focusable
    text={!active ? style.text : style.bg}
    bg={active ? style.text : undefined}
    bg-hocus={style.text}
    text-hocus={style.bg}
    p={{ x: theme.spacing.md, y: theme.spacing.sm }}
    m={{ t: theme.spacing.sm }}
    m-lg={{ t: 0, r: theme.spacing.sm }}
    rounded={theme.radius}
    className={classnames('block no-underline', className)}
    aria-current={active ? 'page' : undefined}
    {...rest}
  >
    {children}
  </BaseComponent>
)

NavItem.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  className: PropTypes.string,
  header: PropTypes.shape({
    style: PropTypes.object.isRequired,
  }),
  active: PropTypes.bool,
}

NavItem.defaultProps = {
  is: 'a',
  children: undefined,
  className: undefined,
  header: {},
  active: false,
}

export { NavItem as component }
export default withTheme(NavItem)
