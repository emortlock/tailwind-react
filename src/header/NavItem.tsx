import type { TouchableProps } from 'tailwind-react-primitives'
import type { WithTheme } from '../theme'
import type { WithHeader } from './Header'

import React from 'react'

import { RawTouchable as Touchable } from 'tailwind-react-primitives'
import { withTheme } from '../theme'

export interface NavItemProps<E extends HTMLElement = HTMLAnchorElement>
  extends TouchableProps<E>,
    WithTheme,
    WithHeader {
  active?: boolean
}

export const RawNavItem = <E extends HTMLElement = HTMLAnchorElement>({
  theme,
  as,
  children,
  header: { style, screen } = { style: {}, screen: 'lg' },
  active,
  ...rest
}: NavItemProps<E>) => {
  const textColor = style?.text || theme.textColors.on.primary
  const bgColor = style?.bg || theme.brandColors.primary

  const responsive = screen
    ? {
        [`m-${screen}`]: { t: 0, r: theme.spacing.sm },
      }
    : {}

  return (
    <Touchable<E>
      as={as}
      focusable
      text={!active ? style?.text : style?.bg}
      bg={active ? textColor : undefined}
      bg-hocus={textColor}
      text-hocus={bgColor}
      p={{ x: theme.spacing.md, y: theme.spacing.sm }}
      m={{ t: theme.spacing.sm }}
      rounded={theme.radius}
      block
      aria-current={active ? 'page' : undefined}
      {...responsive}
      {...rest}
    >
      {children}
    </Touchable>
  )
}

export const NavItem = withTheme(RawNavItem)
