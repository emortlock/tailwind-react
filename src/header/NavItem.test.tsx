import React from 'react'
import { render, RenderResult } from '@testing-library/react'

import { RawNavItem as NavItem } from './NavItem'
import { defaultTheme } from '../theme'

const setup = (
  testProps = {},
): [RenderResult, { props: Record<string, any> }] => {
  const props = Object.assign(
    {
      theme: defaultTheme,
      header: { style: { bg: 'blue-400', text: 'white' } },
    },
    testProps,
  )

  const wrapper = render(<NavItem {...props}>Link</NavItem>)

  return [
    wrapper,
    {
      props,
    },
  ]
}

describe('NavItem', () => {
  it('renders matching snapshot', () => {
    // eslint-disable-next-line
    const [{ asFragment }] = setup({
      as: (props: React.HTMLProps<HTMLAnchorElement>) => (
        // eslint-disable-next-line jsx-a11y/anchor-has-content
        <a {...props} href="#test" />
      ),
    })

    expect(asFragment()).toMatchSnapshot()
  })
})
