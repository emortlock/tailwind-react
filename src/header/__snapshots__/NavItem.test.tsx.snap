// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavItem renders matching snapshot 1`] = `
<DocumentFragment>
  <a
    class="select-none cursor-pointer text-white hover:bg-white focus:bg-white hover:text-blue-400 focus:text-blue-400 px-4 py-2 mt-2 rounded block focus:outline-none focus:shadow-outline"
    href="#test"
    role="button"
    tabindex="0"
  >
    Link
  </a>
</DocumentFragment>
`;
